src/main/haskell/kore/src/Kore/Step/Substitution.hs

```diff
@@ -83,36 +82,41 @@ normalize
     -> StepPatternSimplifier level
     -> BuiltinAndAxiomSimplifierMap level
     -> ExpandedPattern level variable
-    -> Simplifier ( ExpandedPattern level variable )
+    -> BranchT Simplifier (ExpandedPattern level variable)
 normalize
     tools
     substitutionSimplifier
     simplifier
     axiomIdToSimplifier
     Predicated { term, predicate, substitution }
   = do
-    x <- runExceptT $
-        normalizeSubstitutionAfterMerge
+    result <-
+        runExceptT
+        $ normalizeWorker
             tools
             substitutionSimplifier
             simplifier
             axiomIdToSimplifier
             Predicated { term = (), predicate, substitution }
-    return $ case x of
-      Right (Predicated { predicate = p, substitution = s }, _) ->
-          if Predicate.isFalse p
-              then ExpandedPattern.bottom
-              else Predicated term p s
-      Left _ ->
-          Predicated
-              { term
-              , predicate =
-                    makeAndPredicate predicate
-                    $ substitutionToPredicate substitution
-              , substitution = mempty
-              }
+    case result of
+        Right normal
+          | ExpandedPattern.isBottom normal ->
+            return (ExpandedPattern.bottomOf patternSort)
```

Comment:
Actually, this check is redundant because the `\bottom` branches have already been pruned by this point.

Reply:
Speaking of which, OrOfExpandedPattern was responsible for branch pruning. How does that work when using BranchT? Should we check for bottom before each return and perhaps return empty?

Reply:
`scatter` and `gather` operate through `OrOfExpandedPattern`, so we don't need to worry about branch pruning there. New code using `BranchT` should return `empty`. We *do* need to check for bottom when we `lift` `Simplifier a` into `BranchT Simplifier a` that the value of `a` is not bottom, and I will go make sure that all the `lift`s in this pull request do that.

Reply:
It would be nice if it would be impossible to forget to check for bottom when lifting.

Reply:
I.e. would a manual 'MonadTrans' instance solve this?

Also, for new code: we would need to check the pattern we want to return to see if we should use `empty`. As above, this seems too easy to forget, it would be nice if it would work automatically.

btw, nice = not mandatory, but it would probably save us some debugging/some performance loss that would be fairly hard to identify.

Reply:
We can't do this through `MonadTrans`: [`lift`](https://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-Class.html#v:lift) quantifies over the value type. This is a general pattern for all the classes that descend from `Functor`.

I added `liftPruned` and `returnPruned` which do branch pruning on `\bottom`. These make it easy to do the right thing. I would also like to make it hard to do the wrong thing, but I don't know how to do that right now.

Reply:
A compile-time check or something like that might be nice, if possible.